llvm: remove 3.5, 3.8 and 4

lblasc opened the pull request:
###### Motivation for this change
In #66528 we're working on an update of glibc to version 2.30.
LLVM versions [3.5](https://hydra.nixos.org/build/110406734/nixlog/1/tail), [3.8](https://hydra.nixos.org/build/110392337/nixlog/1/tail) and [4](https://hydra.nixos.org/build/110386893/nixlog/1/tail) are very much outdated, they don't build any more. There is similar pull request #77985 for removal of old gcc versions, problematic part is `<sys/ustat.h>` header which was removed in 2.28.

All packages which depended on old llvm version are updated and only pure package is now marked broken.

I'm new to Haskell expressions, hope those changes are fine. cc @peti 
###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

lblasc committed:
dale: 20170519 -> 20181024

lblasc committed:
llvm-general: remove

lblasc committed:
lightspark: 0.8.1 -> 0.8.2

lblasc committed:
haskell.compiler.ghc822Binary: use default llvm version on aarch64

lblasc committed:
haskell.compiler.ghc863Binary: remove llvm from buildInput

Aarch64 binary is not used.

lblasc committed:
pure: mark as broken

lblasc committed:
saw-tools: use default llvm

lblasc committed:
ue4: use default llvm

lblasc committed:
llvm_35: remove

lblasc committed:
configuration-ghc-8.2.x: use default llvmPackages

lblasc committed:
rdedup: use default clang version

lblasc committed:
llvm_39: remove

lblasc committed:
stdenv, darwin: fix setupHook location

lblasc committed:
mozart: build with llvmPackages_5

jchv reviewed (approved):
Lightspark changes LGTM, tested it locally.

lblasc committed:
lutris: use default llvm in chrootenv

lblasc committed:
far2l: build with default llvm

lblasc committed:
llvm_4: remove

peti commented on pkgs/development/haskell-modules/configuration-nix.nix:
```diff
@@ -282,19 +282,20 @@ self: super: builtins.intersectAttrs super {
+  #      librarySystemDepends = [ pkgs.libcxx ] ++ drv.librarySystemDepends or [];
+  #    }
+  #  );
 
```
Please don't comment this out. Rather, please remove the code snippet altogether.

peti commented on pkgs/development/haskell-modules/configuration-ghc-8.2.x.nix:
```diff
@@ -5,7 +5,7 @@ with haskellLib;
 
   # Suitable LLVM version.
-  llvmPackages = pkgs.llvmPackages_39;
+  llvmPackages = pkgs.llvmPackages;
```
This change is fine. The oldest LLVM version we actually need is `llvmPackages_5` for ghc-8.4.4.

ghc-8.2.2 is a binary downloaded from haskell.org that we use for bootstrapping our ghc-8.4.4 and 8.6.5 compilers. We don't need LLVM for that.

lblasc replied:
thx for reviewing, I will leave it then as is.

Ma27 commented:
Is there anything else that prevents us from merging?

worldofpeace commented on pkgs/development/interpreters/pure/default.nix:
```diff
@@ -40,5 +40,6 @@ stdenv.mkDerivation rec {
     platforms = with lib.platforms;
       linux;
     license = lib.licenses.gpl3Plus;
+    broken = true;
```
please comment why it's broken.

asppsa replied:
Pure requires LLVM 3.5 because it's based on the old JIT API that was removed in subsequent versions.

matthewbauer reviewed (commented):
I would prefer to wait for 20.03 branching before removing things to give a little bit more time for people still using these.

Ma27 commented:
I certainly see your point :)
But in that case you'd suggest to delay the glibc update to 20.09 as well, right?

worldofpeace commented:
> I would prefer to wait for 20.03 branching before removing things to give a little bit more time for people still using these.

I'm for sure not the resident expert on these things in this particular PR, but when it comes to 20.03 it seems to shaping up like a lot of package removals.

Also noted that this PR could use some `throws` in `aliases.nix` to point that the packages have been removed (and to use newer versions).

I'm also with @Ma27 with what you think about glibc 2.30.

matthewbauer commented:
> I'm for sure not the resident expert on these things in this particular PR, but when it comes to 20.03 it seems to shaping up like a lot of package removals.

Yeah, I've noticed the gcc removals, so I suppose llvm is fair game to remove too.

> Also noted that this PR could use some throws in aliases.nix to point that the packages have been removed (and to use newer versions).

We probably need a "removals.nix" or similar. At least originally aliases were only supposed to be things that had an obvious replacement. But I do like the idea of a warning that will also tell you the last release of NixOS that included that attr.

matthewbauer reviewed (approved):
Looks good! Note that it really doesn't need to go to staging.

lblasc commented:
Rebased and switched base to master.

worldofpeace commented:
> > I'm for sure not the resident expert on these things in this particular PR, but when it comes to 20.03 it seems to shaping up like a lot of package removals.
> 
> Yeah, I've noticed the gcc removals, so I suppose llvm is fair game to remove too.
> 
> > Also noted that this PR could use some throws in aliases.nix to point that the packages have been removed (and to use newer versions).
> 
> We probably need a "removals.nix" or similar. At least originally aliases were only supposed to be things that had an obvious replacement. But I do like the idea of a warning that will also tell you the last release of NixOS that included that attr.

:rofl: I actually discussed a `removed.nix` in #nixos-dev and took notes of it
* https://gist.github.com/worldofpeace/4c8b238241ae6c79adce1707cddf0544

I should really open an issue..

Ma27 reviewed (approved):
LGTM :+1:

Should we wait for such a remove.nix or can we merge this btw? :)

worldofpeace commented:
> LGTM +1
> 
> Should we wait for such a remove.nix or can we merge this btw? :)

As I suggested, we can just add `throws` in `aliases.nix` like there already is.

worldofpeace reviewed (changes requested):
See above comment.

Ma27 commented:
Fair enough! (Sorry, forgot about that comment 😅)

Ma27 committed:
llvm_{3,4}: re-add attributes with an error message about their removal

With this change, expressions relying on those old LLVM versions will
fail with a meaningful error message.

Ma27 commented:
@worldofpeace re-added those attributes and assigned those to a `throw`-expression. Also rebased onto latest master and added a release-notes entry.

Anything else to add? :)

lblasc commented:
@Ma27 thanks for taking care of this :)

worldofpeace commented:
Rebased because of release note conflict, and I moved the re-added attrs to aliases.nix https://github.com/NixOS/nixpkgs/pull/77989/commits/f73d7259677b508efa5723a64ffbe0ea45634692.
Just waiting on eval.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/january-2020-in-nixos/5771/1

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/january-2020-in-nixos/5771/5
